cpp/src/arrow/c/abi.h

```diff
@@ -228,6 +228,65 @@ struct ArrowDeviceArrayStream {
 
 #endif  // ARROW_C_DEVICE_STREAM_INTERFACE
 
+#ifndef ARROW_C_ASYNC_STREAM_INTERFACE
+#define ARROW_C_ASYNC_STREAM_INTERFACE
+
+// Similar to ArrowDeviceArrayStream, except designed for an asynchronous
+// style of interaction. While ArrowDeviceArrayStream provides producer
+// defined callbacks, this is intended to be created by the consumer instead.
+// The consumer passes this handler to the producer, which in turn uses the
+// callbacks to inform the consumer of events in the stream.
+struct ArrowAsyncDeviceStreamHandler {
+  // Handler for receiving a schema. The passed in stream_schema should be
+  // released or moved by the handler (producer is giving ownership of it to
+  // the handler).
+  //
+  // The `extension_param` argument can be null or can be used by a producer
+  // to pass arbitrary extra information to the consumer (such as total number
+  // of rows, context info, or otherwise).
+  //
+  // Return value: 0 if successful, `errno`-compatible error otherwise
+  int (*on_schema)(struct ArrowAsyncDeviceStreamHandler* self,
+                   struct ArrowSchema* stream_schema, void* extension_param);
+
+  // Handler for receiving an array/record batch. Always called at least once
+  // unless an error is encountered (which would result in calling on_error).
+  // An empty/released array is passed to indicate the end of the stream if no
+  // errors have been encountered.
+  //
+  // The `extension_param` argument can be null or can be used by a producer
+  // to pass arbitrary extra information to the consumer.
+  //
+  // Return value: 0 if successful, `errno`-compatible error otherwise.
+  int (*on_next)(struct ArrowAsyncDeviceStreamHandler* self,
+                 struct ArrowDeviceArray* next, void* extension_param);
```

Comment:
What should happen if the consumer needs to do some lengthy processing of `next` (e.g. maybe the producer is a simple scanner and the consumer is running a complex analysis query).  Should it force the creation of a new thread task (which is annoying because the array is probably in the CPU cache already and we don't want to have to process it on a new thread) or should it do the analysis as part of the call to `on_next` (which is weird because now we are tying up a producer's thread and would that interfere with the production of new data?).

Reply:
I guess the question is how much complexity we want this to have and who should own that complexity. 

As described, the intent was that a consumer doesn't need to handle concurrency by default because a producer should only call the `on_next` when the previous one returns. But in the scenario you describe, where the processing inside of `on_next` might be lengthy, I think it's fine for the consumer to have to make the decision as to whether or not it should create a new thread task (if it wants to handle things in parallel) or do it inside of `on_next` (and potentially tie up the producer, but that's basically just backpressure essentially).

I think this is a good balance and tradeoff of complexity vs ease of use.

Reply:
@zeroshade is easy of use really a concern for such a low-level interface? I think limiting potential is worse than making it hard to user directly. Very sophisticated users should be able to implement advanced concurrency patterns (e.g. handling out-of-order pieces of the stream).

Reply:
IMO, yes, to some degree. A complex spec that few people can feasibly implement won't get used. (Hence my question below about which paradigm will fit actual consumers/producers)

And the scheme described here is pretty much what gRPC does for its callback-based async API which AFAIK has been proved out within Google at least. (Except it's a little more explicit about callback/backpressure: the consumer needs to explicitly request the next message each time. https://github.com/grpc/proposal/blob/master/L67-cpp-callback-api.md)

Reply:
So actually it's probably closer to Weston's proposal below.

Reply:
For some gRPC implementations, tying up the thread is explicitly not recommended so we may not want to encourage that

Reply:
> Very sophisticated users should be able to implement advanced concurrency patterns (e.g. handling out-of-order pieces of the stream).

@felipecrv IMO this API the wrong level to assume out-of-order delivery. Out of order arrival assumes producer's concurrency. If the producer is capable to produce arrays concurrently, and the consumer consume them concurrently or out-of-order, too, it's better to simply create multiple streams and make multiple of consumer's threads/green threads/coroutines/async tasks consume them on the single producer--single consumer basis. It would be *much* less tricky to implement both for consumers and producers, and quite likely even be more performant, too, because there would be no write contention and concurrent queues on the producer's side. Cf. [single writer principle](https://mechanical-sympathy.blogspot.com/2011/09/single-writer-principle.html).

Reply:
@leventov cool. Nice that we can keep C Data simple and not prevent advanced usages like the one you described.

Reply:
Perhaps we should make this SPSC structure more explicit? It would be easy for newcomers to the API to miss that and expect full synchronization then get a nasty surprise when multiple threads push to a single ArrowAsyncDeviceStreamHandler

Reply:
I agree with @leventov that out-of-order delivery is outside the scope of this. 

@bkietz The documentation explicitly states that the callbacks must be called serially, but can be called from different threads.